File: mcp-code-intelligence-python/src/mcp_code_intel/http/ingest_api.py
"""POST /api/memory/ingest HTTP endpoint — HTTP API for external ingest."""

from __future__ import annotations

import json
from typing import Any

from aiohttp import web


def setup_ingest_routes(app: web.Application, engine_provider: Any) -> None:
    """Register ingest API routes on the aiohttp app."""
    app.router.add_post("/api/memory/ingest", _handle_ingest(engine_provider))


def _handle_ingest(engine_provider: Any):
    """Create handler for POST /api/memory/ingest."""

    async def handler(request: web.Request) -> web.Response:
        engine = engine_provider()
        if engine is None:
            return web.json_response(
                {"error": "Memory not initialized"}, status=503
            )
        try:
            body = await request.json()
        except (json.JSONDecodeError, Exception):
            return web.json_response(
                {"error": "Invalid request body"}, status=400
            )
        content = body.get("content", "").strip()
        if not content:
            return web.json_response(
                {"error": "content is required"}, status=400
            )
        summary = (body.get("summary") or "").strip() or content[:120]
        entry_type = (body.get("type") or "").strip() or "CONTEXT"
        tags = body.get("tags", "")
        source = body.get("source")

        entry_id = engine.knowledge.insert_entry(
            content=content,
            summary=summary,
            entry_type=entry_type,
            tier="WORKING",
            source=source,
            tags=tags,
        )
        return web.json_response(
            {"id": entry_id, "type": entry_type, "tier": "WORKING"},
            status=201,
        )

    return handler

File: mcp-code-intelligence-nodejs/src/http/ingest-api.ts
/**
 * POST /api/memory/ingest HTTP endpoint — HTTP API for external ingest.
 */

import type { IncomingMessage, ServerResponse } from 'http';

export interface IngestRequest {
  content: string;
  type?: string;
  source?: string;
  summary?: string;
  tags?: string;
}

export interface IngestResponse {
  id: number;
  type: string;
  tier: string;
}

/** Handle POST /api/memory/ingest. */
export async function handleIngest(
  req: IncomingMessage,
  res: ServerResponse,
  engine: any
): Promise<void> {
  if (!engine) {
    res.writeHead(503);
    res.end(JSON.stringify({ error: 'Memory not initialized' }));
    return;
  }
  const body = await parseBody(req);
  if (!body || !body.content || !body.content.trim()) {
    res.writeHead(400);
    res.end(JSON.stringify({ error: 'content is required' }));
    return;
  }
  const summary = body.summary?.trim() || body.content.slice(0, 120);
  const type = body.type?.trim() || 'CONTEXT';
  const tags = body.tags ?? '';

  const id = engine.knowledge.insert({
    content: body.content,
    summary,
    type,
    tier: 'WORKING',
    source: body.source ?? null,
    tags,
  });

  const response: IngestResponse = { id, type, tier: 'WORKING' };
  res.writeHead(201, { 'Content-Type': 'application/json' });
  res.end(JSON.stringify(response));
}

async function parseBody(req: IncomingMessage): Promise<IngestRequest | null> {
  return new Promise((resolve) => {
    const chunks: Buffer[] = [];
    req.on('data', (chunk: Buffer) => chunks.push(chunk));
    req.on('end', () => {
      try {
        resolve(JSON.parse(Buffer.concat(chunks).toString()));
      } catch {
        resolve(null);
      }
    });
    req.on('error', () => resolve(null));
  });
}
